File: js/server/json.js
////////////////////////////////////////////////////////////////////////////////
/// @brief JavaScript JSON utility functions
///
/// @file
///
/// DISCLAIMER
///
/// Copyright 2010-2011 triagens GmbH, Cologne, Germany
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///
/// Copyright holder is triAGENS GmbH, Cologne, Germany
///
/// @author Dr. Frank Celler
/// @author Copyright 2011, triAGENS GmbH, Cologne, Germany
////////////////////////////////////////////////////////////////////////////////

// -----------------------------------------------------------------------------
// --SECTION--                                                 AvocadoCollection
// -----------------------------------------------------------------------------

////////////////////////////////////////////////////////////////////////////////
/// @addtogroup V8Json
/// @{
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
/// @brief string representation of a collection
////////////////////////////////////////////////////////////////////////////////

AvocadoCollection.prototype.toString = function() {
  var status;

  if (this instanceof AvocadoCollection) {
    status = this.status();

    if (status == 1) {
      return "[new born collection " + JSON.stringify(this._name) + "]";
    }
    else if (status == 2) {
      return "[unloaded collection " + JSON.stringify(this._name) + "]";
    }
    else if (status == 3) {
      return "[collection " + JSON.stringify(this._name) + "]";
    }
    else {
      return "[corrupted collection " + JSON.stringify(this._name) + "]";
    }
  }
  else {
    return "[object]";
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @brief JSON representation of a collection
////////////////////////////////////////////////////////////////////////////////

AvocadoCollection.prototype.toJSON = function() {
  if (this instanceof AvocadoCollection) {
    status = this.status();

    if (status == 1) {
      return "[new born collection " + JSON.stringify(this._name) + "]";
    }
    else if (status == 2) {
      return "[unloaded collection " + JSON.stringify(this._name) + "]";
    }
    else if (status == 3) {
      return "[collection " + JSON.stringify(this._name) + "]";
    }
    else {
      return "[corrupted collection " + JSON.stringify(this._name) + "]";
    }
  }
  else {
    return "[object]";
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @}
////////////////////////////////////////////////////////////////////////////////

// -----------------------------------------------------------------------------
// --SECTION--                                                   AvocadoDatabase
// -----------------------------------------------------------------------------

////////////////////////////////////////////////////////////////////////////////
/// @addtogroup V8Json
/// @{
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
/// @brief string representation of a vocbase
////////////////////////////////////////////////////////////////////////////////

AvocadoDatabase.prototype.toString = function() {
  if (this instanceof AvocadoDatabase) {
    return "[vocbase at " + JSON.stringify(this._path) + "]";
  }
  else {
    return "[object]";
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @brief JSON representation of a vocbase
////////////////////////////////////////////////////////////////////////////////

AvocadoDatabase.prototype.toJSON = function() {
  if (this instanceof AvocadoDatabase) {
    return "[vocbase at " + JSON.stringify(this._path) + "]";
  }
  else {
    return "[object]";
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @}
////////////////////////////////////////////////////////////////////////////////

// -----------------------------------------------------------------------------
// --SECTION--                                                      AvocadoEdges
// -----------------------------------------------------------------------------

////////////////////////////////////////////////////////////////////////////////
/// @addtogroup V8Json
/// @{
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
/// @brief string representation of a vocbase
////////////////////////////////////////////////////////////////////////////////

AvocadoEdges.prototype.toString = function() {
  if (this instanceof AvocadoEdges) {
    return "[edges at " + JSON.stringify(this._path) + "]";
  }
  else {
    return "[object]";
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @brief JSON representation of a vocbase
////////////////////////////////////////////////////////////////////////////////

AvocadoEdges.prototype.toJSON = function() {
  if (this instanceof AvocadoEdges) {
    return "[edges at " + JSON.stringify(this._path) + "]";
  }
  else {
    return "[object]";
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @}
////////////////////////////////////////////////////////////////////////////////

// -----------------------------------------------------------------------------
// --SECTION--                                            AvocadoEdgesCollection
// -----------------------------------------------------------------------------

////////////////////////////////////////////////////////////////////////////////
/// @addtogroup V8Json
/// @{
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
/// @brief string representation of an edges collection
////////////////////////////////////////////////////////////////////////////////

AvocadoEdgesCollection.prototype.toString = function() {
  var status;

  if (this instanceof AvocadoEdgesCollection) {
    status = this.status();

    if (status == 1) {
      return "[new born collection " + JSON.stringify(this._name) + "]";
    }
    else if (status == 2) {
      return "[unloaded collection " + JSON.stringify(this._name) + "]";
    }
    else if (status == 3) {
      return "[collection " + JSON.stringify(this._name) + "]";
    }
    else {
      return "[corrupted collection " + JSON.stringify(this._name) + "]";
    }
  }
  else {
    return "[object]";
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @brief JSON representation of an edges collection
////////////////////////////////////////////////////////////////////////////////

AvocadoEdgesCollection.prototype.toJSON = function() {
  var status;

  if (this instanceof AvocadoEdgesCollection) {
    status = this.status();

    if (status == 1) {
      return "[new born collection " + JSON.stringify(this._name) + "]";
    }
    else if (status == 2) {
      return "[unloaded collection " + JSON.stringify(this._name) + "]";
    }
    else if (status == 3) {
      return "[collection " + JSON.stringify(this._name) + "]";
    }
    else {
      return "[corrupted collection " + JSON.stringify(this._name) + "]";
    }
  }
  else {
    return "[object]";
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @}
////////////////////////////////////////////////////////////////////////////////

// Local Variables:
// mode: outline-minor
// outline-regexp: "^\\(/// @brief\\|/// @addtogroup\\|// --SECTION--\\|/// @page\\|/// @}\\)"
// End:

File: tests/toast/lib/toast_test/suite_result/json.ex
################################################################################
## DISCLAIMER
##
## Copyright 2014-2026 ArangoDB GmbH, Cologne, Germany
## Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
##
## Licensed under the Business Source License 1.1 (the "License");
## you may not use this file except in compliance with the License.
## You may obtain a copy of the License at
##
##     https://github.com/arangodb/arangodb/blob/devel/LICENSE
##
## Unless required by applicable law or agreed to in writing, software
## distributed under the License is distributed on an "AS IS" BASIS,
## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
## See the License for the specific language governing permissions and
## limitations under the License.
##
## Copyright holder is ArangoDB GmbH, Cologne, Germany
################################################################################

defmodule ToastTest.SuiteResult.JSON do
  @moduledoc false

  alias ToastTest.SuiteResult

  @spec write(SuiteResult.t(), Path.t()) :: :ok
  def write(%SuiteResult{} = result, result_dir) do
    tests = SuiteResult.flat_tests(result)

    data = %{
      "suite" => result.suite,
      "started_at" => DateTime.to_iso8601(result.started_at),
      "finished_at" => if(result.finished_at, do: DateTime.to_iso8601(result.finished_at)),
      "duration_us" => result.times_us.run,
      "summary" => build_summary(tests),
      "tests" => Enum.map(tests, &encode_test/1)
    }

    json = data |> :json.format(&json_encoder/3, %{}) |> IO.iodata_to_binary()
    File.write!(Path.join(result_dir, "outcomes.json"), json)
    :ok
  end

  defp build_summary(tests) do
    counts = Enum.frequencies_by(tests, & &1.outcome)

    %{
      "passed" => Map.get(counts, :passed, 0),
      "failed" => Map.get(counts, :failed, 0),
      "skipped" => Map.get(counts, :skipped, 0),
      "excluded" => Map.get(counts, :excluded, 0),
      "invalid" => Map.get(counts, :invalid, 0),
      "invalidated" => Map.get(counts, :invalidated, 0)
    }
  end

  defp encode_test(test) do
    %{
      "module" => ToastTest.Formatting.display_module_name(test.module),
      "name" => Atom.to_string(test.name),
      "outcome" => Atom.to_string(test.outcome),
      "duration_us" => test.duration_us
    }
  end

  defp json_encoder(nil, _encoder, _opts), do: "null"

  defp json_encoder(value, encoder, opts) when is_atom(value) and not is_boolean(value) do
    value |> Atom.to_string() |> encoder.(encoder, opts)
  end

  defp json_encoder(value, encoder, opts), do: :json.format_value(value, encoder, opts)
end
